### PR TITLE
feat(auth): add simple CSRF auth flow

### LINKED ISSUE #32: Add simple GET-based CSRF Multi-Step Auth flow to BC
A simple multi-step flow where the user must do:

1. `GET` to the login page (i.e `GET /auth/simple-csrf-flow`) to get a csrf token from the response
2. `POST` to the auth API endpoint with that token and valid credentials

### PR DESCRIPTION
**Pull request by alexandrmiee**
closes #32

### CODE REVIEW

File: src/users/users.service.ts [hunk 1]
```diff
@@ -9,8 +9,8 @@ export class UsersService {
   public static readonly LDAP_SEARCH_QUERY = (email) =>
     `(&(objectClass=person)(objectClass=user)(email=${email}))`;
   public static readonly LDAP_ERROR_RESPONSE = `
-      Lookup failed: javax.naming.NamingException: 
-      [LDAP: error code 1 - 000004DC: Lda pErr: DSID-0C0906DC, comment: context not found., data 0, v1db1 ]; 
+      Lookup failed: javax.naming.NamingException:
+      [LDAP: error code 1 - 000004DC: Lda pErr: DSID-0C0906DC, comment: context not found., data 0, v1db1 ];
       remaining name: 'OU=Users,O=BrokenCrystals'
```
Comment (derevnjuk) [inline src/users/users.service.ts:14]: Revert useless changes

File: src/users/users.service.ts [hunk 2]
```diff
@@ -56,6 +56,20 @@ export class UsersService {
...
+    const user = await this.findByEmail(email);
+    if (!user) {
+      throw new Error('Could not find user');
+    }
+    wrap(user).assign({
+      crsfToken: token,
+    });
+
+    await this.usersRepository.persistAndFlush(user);
+    return user;
```
Comment (derevnjuk) [inline src/users/users.service.ts:70]: Use double submit cookie approach rather than synchronizer pattern.

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -60,6 +92,16 @@ export class AuthController {
...
 
+    if (req.csrf) {
+      const token = this.getCsrfTokenFromCookies(cookie);
+      if (token !== req.csrf) {
+        throw new UnauthorizedException({
+          error: 'Invalid credentials',
+          location: __filename,
+        });
+      }
+    }
```
Comment (derevnjuk) [inline src/auth/auth.controller.ts:103]: `login` is not responsible for CSRF token verification. You should move it to a separate middleware

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/auth/auth.controller.ts (hunk 3 above)
Comment (alexandrmiee) [reply src/auth/auth.controller.ts]: In middleware we have no access to body so I moved it to guard

**Comment by alexandrmiee** [comment]
@derevnjuk 
If user select `Simple CSRF-based Authentication` backend will set `_csrf` cookie. But if user wand to change authentication type cookie will stay.
On backend side login will always fail with 401 error if client requests `/login` with `_csrf` cookie. On the other side, we cannot skip checks if csrf field not exists in body.
So can we use separate url to validating cookie?

**Comment by derevnjuk** [comment]
@alexandrmiee why don't use `op` parameter to determine expected behavior as before?
/cc @BMNTeam

**Comment by alexandrmiee** [comment]
`op` can be omitted, but the ` csrf` check is required

**Comment by derevnjuk** [comment]
@alexandrmiee it's the opposite, you should check `csrf` only if corresponding `op` is selected.

File: src/auth/csrf.guard.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    const request: FastifyRequest = context.switchToHttp().getRequest();
+
+    try {
+      const body: LoginRequest = request.body as LoginRequest;
+      if (body?.op === FormMode.CSRF) {
+        const { csrf: csrfToken } = body;
+        const { cookie } = request.headers;
+        let token: string;
+        const cookies: string[] = cookie.split('; ');
```
Comment (derevnjuk) [inline src/auth/csrf.guard.ts:27]: Don't parse `cookies` manually, use `res.cookies` instead

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/auth/csrf.guard.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      if (body?.op === FormMode.CSRF) {
+        const { csrf: csrfToken } = body;
+        const csrfCookie = request.cookies[CsrfGuard.CSRF_COOKIE_HEADER];
+        if (csrfCookie) {
+          const token = decodeURIComponent(csrfCookie);
+          if (token === csrfToken) {
+            return true;
+          }
+        }
+        this.throwError();
```
Comment (derevnjuk) [inline src/auth/csrf.guard.ts:32]: Please refactor this code to reduce its complexity (there are too many branches)

**Review by derevnjuk — CHANGES_REQUESTED** [review]

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**